### ISSUE TITLE
Darkrift 2.4.4 cached messages having wrong data

### ISSUE THREAD

**Issue by Armynator** [DarkRiftNetworking/DarkRift#82]
It looks like incoming messages on a client are not recycled correctly. Sending multiple different messages to a client lets the client use the same outdated messages over and over instead of receiving the correct data.

First I'm connecting with a Unity Client to a Unity Server to send an initial request like this:
```
private void Start() {
    if (client.ConnectionState == ConnectionState.Connected) {
        client.MessageReceived += ServerMessageReceived;

        using (DarkRiftWriter writer = DarkRiftWriter.Create()) {
            using (Message message = Message.Create(1, writer)) {
                client.SendMessage(message, SendMode.Reliable);
            }
        }
    }
}

private void ServerMessageReceived(object sender, MessageReceivedEventArgs e) {
    using (Message message = e.GetMessage() as Message) {
        Debug.Log(message.Tag);
    }
}
```

The server receives the request correctly, so I'm sending back 20 messages only containing a different tag every time:
```
void ClientConnected(object sender, ClientConnectedEventArgs e) {
    Debug.Log("Connected: " + e.Client.ID);
    e.Client.MessageReceived += HandleClientRequest;
}

private void HandleClientRequest(object sender, MessageReceivedEventArgs e) {
    for (int i = 0; i < 20; i++) {
        using (DarkRiftWriter writer = DarkRiftWriter.Create()) {
            using (Message messagee = Message.Create((ushort)i, writer)) {
                e.Client.SendMessage(messagee, SendMode.Reliable);
	    }
        }
    }
}
```

With the default cache settings, I'm getting wrong results: 
![1](https://user-images.githubusercontent.com/5256247/59958772-e98a8a80-94ab-11e9-888d-3158cb173ba4.png)
As you can see, the same tag is received multiple times, and the order is totally wrong.

However, once setting max writers, readers, messages and message buffers to 0 in the object cache settings, the messages are received correctly:
![2](https://user-images.githubusercontent.com/5256247/59958791-69b0f000-94ac-11e9-9d66-49e368fcf428.png)

Also note that I disabled caching in the server config as well. With caching enabled, the server isn't sending the messages in the correct order. Not sure if that's intended, but thats probably a different issue anyway.

Edit 23.06.2019:
Tested with 2.4.4, issue still persists. Only messages and message buffers seem to be affected. Writers and readers are fine.
Tested with code on client:

```
using (Message message = e.GetMessage() as Message) {
            using (DarkRiftReader reader = message.GetReader()) {
                Debug.Log(message.Tag + " - " + reader.ReadInt32());
            }
}
```

Server:
```
for(int i = 0; i < 20; i++) {
    using (DarkRiftWriter writer = DarkRiftWriter.Create()) {
        writer.Write(i);
        using (Message messagee = Message.Create((ushort)i, writer)) {
            messagee.Serialize(writer);
            e.Client.SendMessage(messagee, SendMode.Reliable);
        }
    }
}
```

Results:
![r1](https://user-images.githubusercontent.com/5256247/59971475-fd51f180-957c-11e9-966b-dd1e08f53f32.png)
Everything fine with max messages and max message buffers set to 0. (Except the order being not correct)

![r2](https://user-images.githubusercontent.com/5256247/59971484-1f4b7400-957d-11e9-8ec9-31fbaf2c75ce.png)
Once you allow messages to cache you get the same message tag over and over, and even incorrect data from the reader.

**Comment by JamJar00** [comment]
Hi Armynator,

Firstly, thanks for a thorough and well written issue: It makes my life a lot easier!

Some of what you're seeing is expected behaviour and other things are (very obviously) bugs. DarkRift doesn't guarantee order at the moment because the internal threading that DarkRift does basically means it would have to reorder the messages again before passing them on to plugin code. This is something I'm hoping to address in DarkRift 3.

The reason that disabling the caches gives the illusion of ordered messages is because every time DarkRift calls to the .NET framework to create an object it has to access a global memory lock which slows everything down and causes all the threads to progress together in a sort of rough lockstepped motion as they request new objects. When the cache is enabled they rarely have to aquire that lock as they instead use their thread local cache to get the new objects.

Unfortunately however, there are bugs currently causing the objects to seep between the threads somehow which leads to the behaviour you see in the final screenshots where messages are corrupted and duplicated quite frequently. This is still an issue even with cache disabled but the effects are much less frequenty seen due to the constant synchronization against that global lock.

It's good to keep getting examples of these to monitor the extent of the issue so thanks for reporting this. I'm continuing to work on finding these bugs as my top priority and don't plan on releasing 2.5.x releases until this is stabilised.

Let me know if you have any other info or questions and thanks again for raising the issue.

Jamie

**Comment by JamJar00** [comment]
Hi Armynator,

I'm hoping that 2.4.5 has fixed this issue, would you be able to give it a go in your project and report back as to whether there are still any issues or not please?

You can download 2.4.5 here:
https://www.darkriftnetworking.com/DarkRift2/Releases/Free/2.4.5.unitypackage
https://www.darkriftnetworking.com/DarkRift2/Releases/Pro/2.4.5.unitypackage?invoice=<your invoice number>

Jamie

**Comment by Armynator** [comment]
Fixed in 2.4.5 :)
